app/controllers/sessions_controller.rb

```diff
@@ -9,6 +10,31 @@ def create
     redirect_to origin || user_path
   end
 
+  def new_or_existing(auth)
+    user = user_for_identity(auth)
+    return create_user_and_identity(auth) if user.blank?
+
+    user.omniauth_tokens(auth)
+    return user
+  end
+
+  def user_for_identity(auth)
+    identities = Identity.where(provider: auth.provider, uid: auth.uid)
+
+    return unless identities.any?
+
+    identities.first&.user
+  end
```

Comment:
It would be simpler to just look the user up directly by email address, but it's not safe: we want users to be able to edit their email address separately from their social media email address, so it might not match